### PR TITLE
Proposal for simpler handling of unsafe sig and fix for #691

### PR DESCRIPTION
**Pull request by jamescheney**
This PR disables the previous code for handling unsafe signatures and implements a new approach:

- `unsafe sig foo` means that the declared identifier is not treated as recursive, for the purpose of effect typing; that is, recursive calls to `foo` are tame.
- type inference is otherwise as normal, and we do not (and do not need to) adjust the inferred type

It also adds treatment of unsafe-signature behavior to the IR.  This unfortunately touches a lot of code because the function definition data is a tuple rather than record.  In IR typechecking, we treat identifiers in recursive blocks as non-recursive if they are declared with `unsafe sig`.

Finally, a remaining problem is that pattern matching against List cons is translated (by `compilePatterns.ml`) to calls to `hd` and `tl`, which are wild.  This PR introduces a small hack to work around this: adding tame versions of `hd` and `tl` which have names that can never appear in Links source code, and are only used for desugaring of pattern matching.  This is slightly bad though because it would mean that these versions of `hd` and `tl` could be present in query code and the IR typechecker would allow this; but this is a problem we already have because of #37.  I have some longer-term ideas how to remove this related to #47 and #37.

### CODE REVIEW

**Comment by jamescheney** [comment]
Mental note: before merging update typed IR tests that currently have dummy code for functions with unsafe sig.

**Comment by frank-emrich** [comment]
This sounds like a reasonable approach, given that it makes it much less invaisve to handle unsafe signatures in the IR type-checker.
I've had a look at the changes to `IrCheck`, and they look good to me.

**Comment by jamescheney** [comment]
The following six ways of defining `hd` typecheck (and are tame) in Links:
```
fun ([x]) {x};
fun (x::_) {x};
fun (l) {var [x] = l; x};
fun (l) {var x::_ = l; x};
fun (l) {switch (l) {case [x] -> x}};
fun (l) {switch (l) {case x::_ -> x}};
```
However, in this branch, the first two don't pass the IR typechecker, while the remaining ones do.  This means that an occurence of the first one in `prelude.links` `formP` also fails to IR typecheck.  

The problem was that for functions (and only functions) the ambient environment of the function expression is not the one you want to use for the desugared pattern code.  Thus, when translating to IR, the wrong effect (e.g. wild) was being given as a type argument to the effect-polymorphic occurrences of `hd` and `tl` generated by the pattern compiler.

The fix is to use the inner environment (body_env) when translating the desugared patterns to IR, not the outer environment (which might have a different ambient effect, incompatible with the inner one).

**Comment by jamescheney** [comment]
I'm really tempted to make `Ir.fun_def` into a record though, since this would pretty much touch all the same code.

**Comment by jamescheney** [comment]
Just realized that I would have saved myself some trouble if I'd reread the comments and seen that @jstolarek already did some of the work that I just redid in a branch 
`js-T691-ir-unsafe-sigs`
Of course, I made different, incompatible  arbitrary decisions.

File: core/typeSugar.ml [hunk 1]
```diff
@@ -362,7 +362,7 @@ sig
   val bind_rec_annotation : griper
   val bind_rec_rec : griper
 
-  val bind_unsafe_fun_annotation : griper
+(*XXX  val bind_unsafe_fun_annotation : griper *)
```
Comment (jstolarek) [inline core/typeSugar.ml:365]: Why leave this?

Comment (jamescheney) [reply core/typeSugar.ml:365]: I meant to remove it and missed it.  Thanks.

**Comment by jamescheney** [comment]
> What exactly happens? Do these definitions result in one of the known IR errors (there shouldn't be any by now except for #865?) or some new IR typing error? Either way I think this should be recorded, either as a new bug report or a comment on existing bug report. 


It's basically an effect-mismatch type error.  It wasn't a bug in the IR typechecker, instead, the IR typechecker was correctly catching the fact that earlier stages were putting incorrect `wild` annotations on the code.

To illustrate what was happening in a hopefully somewhat readable way...

We compile desugared Links code with patterns in function parameters by, first, doing a little extra desugaring.  (Which could perhaps be hoisted out as a separate desugaring step, but currently happens in tandem with sugartoir/compilePatterns).  A function with patterns in its parameters like
```
fun ([x]) {x}
```
gets desugared to
```
fun (z) {var [x] = z; x}
```
Then this in turn gets pattern-compiled to some IR code along these lines:
```
fun (z) { 
  if (empty(x) )
   {Wrong} 
  else { 
    var hx = ##hd(z); 
    var tz = ##tl(z); 
    if (empty(tz)) { var x = hx; x }
    else {Wrong}
  }
```
But wait, you may ask, that Links code you just wrote isn't what we're really working with, we're working with desugared, typechecked ASTs and we need to maintain the type information.  Indeed we do.  Also, `##hd` and `##tl` are tame (i.e. fully effect-polymorphic) versions of `hd` and `tl`, so we need to plug in an effect for them.  What effect should we use?  

The right answer, of course, is the effect that we have already inferred for the body of the (in this case, anonymous) function, because this should match whatever the rest of the function does.  However, what was happening, possibly due to a typo, was that the ambient effect that has been inferred for code *outside* of the function was being used instead - in this case, since the function was checked at the top level, this would be `wild`.  The solution, done in 581ca95b4dbcc59df867b4237abf180a09a3aa31, is to modify `sugartoir` so that the correct environment is indeed used.  Because of the way the translation to IR was defined, however, for recursive functions this requires further refactoring because *the effects inferred for each function in a mutual block might be different* (though they all might be wild, some might have other effects, or be polymorphic over different effect variables bound in different places).

I thought it would be a little silly to make an issue about this and immediately close it, but I can if you want - that would help document this.  Or maybe this is enough.

> Also, given that there are alternative ways of defining hd do we need the ##hd hack? Couldn't we redefine hd in prelude instead?

We could, but I wouldn't - it's a longstanding bug that `hd`/`tl` can be emulated in queries using pattern matching (#37) and one way of fixing this is to make pattern-matching against list `cons` attract the wild effect so we would wind up back where we started.

> 
> I found the explanations in the first comment of the PR helpful - could they make it into the code as a Note somewhere?

OK
- [ ] TODO: add explanatory note about unsafe sig

**Comment by jstolarek** [comment]
> I thought it would be a little silly to make an issue about this and immediately close it, but I can if you want - that would help document this.

If this is fixed as part of this PR then it doesn't make sense to make a separate bug report - disregard my earlier comment.

**Pull request merged**